simulariumio/mcell/mcell_converter.py

```diff
@@ -0,0 +1,347 @@
+        """
+        Get euler angles in degrees representing a rotation defined by the basis
+        between the given normal and a perpendicular vector rotated at angle
+        """
+        perpendicular = McellConverter.get_perpendicular_vector(normal, angle)
+        rotation = McellConverter.get_rotation_matrix(normal, perpendicular)
+        return Rotation.from_matrix(rotation).as_euler("xyz", degrees=True)
+
+    @staticmethod
+    def _get_rotation_euler_angles_for_normals(
+        normals: np.ndarray, angle: float = None
+    ) -> np.ndarray:
+        """
+        Generate an orientation around each normal and return euler angles
+        Either use the given angle or random ones
+        """
+        if angle is None:
+            angles = np.rad2deg(2 * np.pi) * np.random.random(normals.shape[0])
+        else:
+            angles = np.array(normals.shape[0] * [angle])
+        return np.array(
+            [McellConverter.get_euler_angles(n, a) for n, a in zip(normals, angles)]
+        )
+
+    @staticmethod
+    def _should_read_cellblender_binary_file(
+        file_name: str, nth_timestep_to_read: int
+    ) -> bool:
+        """
+        Is this a cellblender binary file?
+        And should this frame of data be included in the visualization?
+        """
+        if "ascii" in file_name or not file_name.endswith(".dat"):
+            return False
+        split_file_name = file_name.split(".")
+        time_index = int(split_file_name[split_file_name.index("dat") - 1])
+        return time_index % nth_timestep_to_read == 0
+
+    @staticmethod
+    def _count_agents_in_binary_cellblender_viz_frame(file_name: str) -> int:
+        """
+        Count the number of agents in the frame of cellblender data
+        """
+        total_mols = 0
+        with open(file_name, "rb") as mol_file:
+            # first 4 bytes must contain value '1'
+            b = array.array("I")
+            b.fromfile(mol_file, 1)
+            assert b[0] == 1
+            while True:
+                try:
+                    # advance through file to get
+                    # number of instances of each molecule type
+                    n_chars_type_name = array.array("B")
+                    n_chars_type_name.fromfile(mol_file, 1)
+                    mol_file.seek(n_chars_type_name[0], os.SEEK_CUR)
+                    is_surface_mol = array.array("B")
+                    is_surface_mol.fromfile(mol_file, 1)
+                    is_surface_mol = is_surface_mol[0] == 1
+                    n_data = array.array("I")
+                    n_data.fromfile(mol_file, 1)
+                    n_data = n_data[0]
+                    data = array.array("f")
+                    data.fromfile(mol_file, n_data)
+                    if is_surface_mol:
+                        data.fromfile(mol_file, n_data)
+                    total_mols += int(n_data / 3.0)
+                except EOFError:
```

Comment:
this EOFError check seems clunky - the file has no indication of how many times you'll be able to run this loop?

Reply:
yeah, the binary data doesn't include that https://cnl.salk.edu/~ahusar/mcell4_documentation/file_formats.html#molecule-visualization-data and this is how they implemented reading it 🤷‍♀️